src/hooks/useConfirmation/index.tsx

```diff
@@ -0,0 +1,166 @@
+import Modal, { Props as ModalProps } from '../../components/Modal';
+import Button, { Props as ButtonProps } from '../../components/Button';
+
+function removeElementsWithUndefinedValue<T extends Record<string, any>>(obj: T | undefined) {
+    if (obj === undefined) {
+        return obj;
+    }
+
+    const newObj = {...obj};
+
+    Object.keys(newObj).forEach((key) => {
+        if (obj[key] === undefined) {
+            delete newObj[key];
+        }
+    });
+
+    return newObj;
+}
+
+
+export interface Options {
+    showConfirmationInitially: boolean;
+    onConfirm?: () => void;
+    onDeny?: () => void;
+    onResolve?: (hasConfirmed: boolean) => void;
+
+    heading?: React.ReactNode;
+    message?: React.ReactNode;
+
+    denyButtonIcons?: React.ReactNode;
+    denyButtonContent?: React.ReactNode;
+    denyButtonActions?: React.ReactNode;
+    denyButtonClassName?: string;
+
+    confirmButtonIcons?: React.ReactNode;
+    confirmButtonContent?: React.ReactNode;
+    confirmButtonActions?: React.ReactNode;
+    confirmButtonClassName?: string;
+}
+
+export type ExtraProps = Omit<ModalProps, 'onCloseButtonClick' | 'heading'>;
+
+const defaultOptions: Partial<Options> = {
+    message: 'Are you sure?',
+    heading: 'Confirmation',
+    confirmButtonContent: 'Yes',
+    denyButtonContent: 'No',
+};
+
+function useConfirmation(options?: Options, extraProps?: ExtraProps): [
+    React.ReactNode,
+    ButtonProps<string | undefined>['onClick'],
+] {
+    const {
+        showConfirmationInitially = false,
+        onConfirm,
+        onDeny,
+        onResolve,
+        heading,
+        message,
+        denyButtonActions,
+        denyButtonContent,
+        denyButtonIcons,
+        denyButtonClassName,
+        confirmButtonActions,
+        confirmButtonContent,
+        confirmButtonIcons,
+        confirmButtonClassName,
+    } = {
+        ...defaultOptions,
+        ...removeElementsWithUndefinedValue(options),
+    };
+
+    const [showModal, setShowModalTrue, setShowModalFalse] = useBooleanState(showConfirmationInitially);
+
+    const handleCancelButtonClick = React.useCallback(() => {
+        setShowModalFalse();
+        if (onDeny) {
+            onDeny();
+        }
+
+        if (onResolve) {
+            onResolve(false);
+        }
+    }, [setShowModalFalse]);
+
+    const handleConfirmButtonClick = React.useCallback(() => {
+        setShowModalFalse();
+        if (onConfirm) {
+            onConfirm();
+        }
+
+        if (onResolve) {
+            onResolve(true);
+        }
+    }, [setShowModalFalse]);
+
+    const modal = React.useMemo(() => {
+        if (!showModal) {
+            return null;
+        }
+
+        const {
+            hideCloseButton,
+            footerActions,
+            ...modalProps
+        } = extraProps ?? {};
+
+        return (
+            <Modal
+                {...modalProps}
+                heading={heading}
+                onCloseButtonClick={handleCancelButtonClick}
```

Comment:
For resolution, I don't think closing the popup is a cancel action. It might be I don't want to deal with this action.
So, for close button, lets just close the popup instead of resolving it as a cancel.
What do you think?

Reply:
But on second thought, we need to have a decision on the popup, so maybe lets not have a close button?
Or should we not think too much? :yum:

Reply:
So what if you want to confirm before closing the modal, in case there's some important form inside the modal?